assignment-0/pkg/controller/controller.go

```diff
@@ -0,0 +1,155 @@
+	"k8s.io/client-go/util/workqueue"
+)
+
+type Controller struct {
+	Client   *kubernetes.Clientset
+	Informer cache.SharedIndexInformer
+	Queue    workqueue.RateLimitingInterface
+	// ServiceDelete if set to false, prevents service deletion on controller restart
+	ServiceDelete bool
+}
+
+func NewController(client *kubernetes.Clientset, informerFactory informers.SharedInformerFactory) *Controller {
+	deploymentInformer := informerFactory.Apps().V1().Deployments().Informer()
+	c := &Controller{
+		Client:   client,
+		Informer: deploymentInformer,
+		Queue:    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "myqueue"),
+	}
+	deploymentInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.handleAdd,
+		DeleteFunc: c.handleDelete,
+	})
+	return c
+}
+
+func (c *Controller) Run(ch <-chan struct{}) {
+	fmt.Println("starting controller")
+	if !cache.WaitForCacheSync(ch, c.Informer.HasSynced) {
+		fmt.Print("waiting for cache to be synced\n")
+	}
+
+	go wait.Until(c.worker, 1*time.Second, ch)
+
+	<-ch
+}
+
+func (c *Controller) worker() {
+	for c.processItems() {
+	}
+}
+
+func (c *Controller) getServiceName(deployment *appsv1.Deployment) string {
+	return deployment.Name + "-service"
+}
+
+func (c *Controller) processItems() bool {
+	item, shutdown := c.Queue.Get()
+	if shutdown {
+		return false
+	}
+	c.Queue.Forget(item)
+	deployment := item.(*appsv1.Deployment)
+	serviceName := c.getServiceName(deployment)
+	service, err := c.Client.CoreV1().Services(deployment.Namespace).Get(
+		context.Background(), serviceName, metav1.GetOptions{})
+	if err == nil {
+		if c.ServiceDelete {
+			serviceError := c.deleteService(service)
+			if serviceError != nil {
+				fmt.Println(serviceError.Error())
+				return false
+			}
+			fmt.Printf("Service %s deleted", service.Name)
+		}
+	} else {
+		if !c.ServiceDelete {
+			serviceError := c.createService(deployment)
+			if serviceError != nil {
+				fmt.Println(serviceError.Error())
+				return false
+			}
+			fmt.Printf("Created service for deployment %s\n", deployment.Name)
+		}
+	}
+	return true
+}
+
+func (c *Controller) handleAdd(obj interface{}) {
+	fmt.Println("Handling deployment add")
+	c.Queue.Add(obj)
+	c.ServiceDelete = false
+}
+
+func (c *Controller) handleDelete(obj interface{}) {
+	fmt.Println("Handling deployment delete")
+	c.Queue.Add(obj)
+	c.ServiceDelete = true
+}
+
+func (c *Controller) getServicePorts(deployment *appsv1.Deployment) []corev1.ServicePort {
+	servicePorts := []corev1.ServicePort{}
+	for _, container := range deployment.Spec.Template.Spec.Containers {
+		for _, containerPort := range container.Ports {
+			port := corev1.ServicePort{
+				Name:     containerPort.Name,
+				Protocol: containerPort.Protocol,
+				Port:     containerPort.ContainerPort,
+			}
+			servicePorts = append(servicePorts, port)
+		}
+	}
+	return servicePorts
+}
+
+func (c *Controller) createServiceInKubernetes(deployment *appsv1.Deployment) error {
```

Comment:
this should be just `createService` I think. It's implicit that we are going to create the service in Kubernetes.

Reply:
I have another function with same name. hence used `createServiceInKubernetes` as a name